src/api-binder.ts

```diff
@@ -0,0 +1,888 @@
+			apiPrefix: baseUrl,
+			passthrough,
+		});
+		this.cachedBalenaApi = this.balenaApi.clone({}, { cache: {} });
+	}
+
+	public async loadBackupFromMigration(retryDelay: number): Promise<void> {
+		try {
+			const exists = await pathExistsOnHost(
+				Path.join('mnt/data', constants.migrationBackupFile),
+			);
+			if (!exists) {
+				return;
+			}
+			console.log('Migration backup detected');
+			const targetState = await this.getTargetState();
+			await this.deviceState.restoreBackup(targetState);
+		} catch (err) {
+			console.log('Error restoring migration backup, retrying: ', err);
+
+			return Bluebird.delay(retryDelay).then(() => {
+				return this.loadBackupFromMigration(retryDelay);
+			});
+		}
+	}
+
+	public async start() {
+		const conf = await this.config.getMany([
+			'apiEndpoint',
+			'unmanaged',
+			'bootstrapRetryDelay',
+		]);
+		let { apiEndpoint } = conf;
+		const { unmanaged, bootstrapRetryDelay } = conf;
+
+		if (unmanaged) {
+			console.log('Unmanaged mode is set, skipping API binder initialization');
+			// If we are offline because there is no apiEndpoint, there's a chance
+			// we've went through a deprovision. We need to set the initialConfigReported
+			// value to '', to ensure that when we do re-provision, we'll report
+			// the config and hardward-specific options won't be lost
+			if (!apiEndpoint) {
+				await this.config.set({ initialConfigReported: '' });
+			}
+			return;
+		}
+
+		console.log('Ensuring device is provisioned');
+		await this.provisionDevice();
+		const conf2 = await this.config.getMany([
+			'initialConfigReported',
+			'apiEndpoint',
+		]);
+		apiEndpoint = conf2.apiEndpoint;
+		const { initialConfigReported } = conf2;
+
+		// Either we haven't reported our initial config or we've been re-provisioned
+		if (apiEndpoint !== initialConfigReported) {
+			console.log('Reporting initial configuration');
+			await this.reportInitialConfig(
+				apiEndpoint as string,
+				bootstrapRetryDelay as number,
+			);
+		}
+
+		console.log('Starting current state report');
+		await this.startCurrentStateReport();
+
+		await this.loadBackupFromMigration(bootstrapRetryDelay as number);
+
+		this.readyForUpdates = true;
+		console.log('Starting target state poll');
+		this.startTargetStatePoll();
+	}
+
+	public async fetchDevice(
+		uuid: string,
+		apiKey: string,
+		timeout: number,
+	): Promise<Device | null> {
+		const reqOpts = {
+			resource: 'device',
+			options: {
+				$filter: {
+					uuid,
+				},
+			},
+			passthrough: {
+				headers: {
+					Authorization: `Bearer ${apiKey}`,
+				},
+			},
+		};
+
+		if (this.balenaApi == null) {
+			throw new InternalInconsistencyError(
+				'fetchDevice called without an initialized API client',
+			);
+		}
+
+		try {
+			// TODO: Maybe type this response?
+			const res = (await this.balenaApi
+				.get(reqOpts)
+				.timeout(timeout)) as Device[];
+			return res[0];
+		} catch (e) {
+			return Bluebird.resolve(null);
```

Comment:
This can be `return null` because we're in an async function so it'll automatically be wrapped